doc/src/gotchas.rst

```diff
@@ -784,3 +784,65 @@ from SymPy 0.6.5:
         """
         expr = Poly.cancel(powsimp(expr))
         return powsimp(together(expr.expand()), combine='exp', deep=True)
+
+
+
+Be careful about substituting for derivatives! 
+==============================================
+
+    >>> import sympy as sm
+    >>> import sympy.physics.mechanics as me
+    >>> q1, q2 = me.dynamicsymbols('q1, q2')
+    >>> a, b, c = sm.symbols('a, b, c')
+    >>> expr = q1.diff().diff()**2*q1 + q1.diff()*q2.diff()
+    >>> expr
+    q1(t)*Derivative(q1(t), (t, 2))**2 + Derivative(q1(t), t)*Derivative(q2(t), t)
+     
+.. math::
+    \displaystyle \operatorname{q_{1}}{\left(t \right)} \left(\frac{d^{2}}{d t^{2}} \operatorname{q_{1}}{\left(t \right)}\right)^{2} + \frac{d}{d t} \operatorname{q_{1}}{\left(t \right)} \frac{d}{d t} \operatorname{q_{2}}{\left(t \right)}
+
+It is worth looking at the raw form (the non-rendered form) to see what is going on:
+
+    >>> repr(expr)
+    'q1(t)*Derivative(q1(t), (t, 2))**2 + Derivative(q1(t), t)*Derivative(q2(t), t)'
+
+We are going to be substiting expressions for :math:`q`'s, :math:`\dot{q}`'s, and :math:`\dot{u}`'s. There are a couple of key things to be aware of:
+
+    * subs() and xreplace() do not substitute the expression in a fixed order (i.e. Python dictionaries are not ordered), so if you pass in more than one item in a single subs() call, things might get substituted in an unexpected order.
```

Comment:
In this case `subs` orders the items. Rather than accept this as a `gotcha` perhaps you could investigate using `ordered` rather than the more naive `count_ops/length of args` approach that is used in subs. Notice how `ordered` gets the order of keys correct for this example:
```python
>>> reps={q1: a, q1.diff(): a*b*c, q1.diff().diff(): d}
>>> list(ordered(reps))
[q1(t), Derivative(q1(t), t), Derivative(q1(t), (t, 2))]
```
You could try changing all the code in `subs` after the `if unordered:` to
```python
        if unordered:
            sequence = list(reversed(ordered(sequence)))
```

Reply:
A diff toward that end is included on the issue page #17823.

Reply:
Thanks for reviewing @smichr. I will change the code in basic.py. Is there anything else that needs to be done?

Reply:
We'll see after the full tests run. It's a good sign that `subs` and `core` passed. Also, I think the note in gotcha can be removed. The only note I think that could be added is that for the case of doing subs where the new expression contains a symbol that is being replaced require an ordered sequence (in general) or an application of `repsort` (or equivalent). That's really an advanced substitution topic, however.

As I see it there subs gets progressively more complex in this order:
* single substitution [(x, y)]
* multiple atomic unrelated substitutions [(x, 1), (y, 2)]
* swapping atomic subs [(x, y), (y, x)] - use simultaneous=True
* multiple old-related substitution [(f(x), x**2), (f(x).diff(x), g(x))]
* multiple "topographically related" subs [(x, z), (y, x)] - use repsort

Reply:
Note: the diff in the issue has the working code, not the simple suggestion above

Reply:
Okay. I think i should close this pull request and open up a new one with the working code.

Reply:
no need -- this will update when you push the changes here

Reply:
a change in title and a release note indicate that a unordered substitutions are now sorted with ordered and those that are not atomic (like symbols) are substituted in reverse order e.g. `g(x,y) before f(x) before x before y`.

Writing that makes me wonder if `ordered` should only be used to break ties when the number of args is the same. Let's see how testing goes.